scalac-scoverage-plugin/src/main/scala/scoverage/plugin.scala

```diff
@@ -107,10 +130,47 @@ class ScoverageInstrumentationComponent(val global: Global, extraAfterPhase: Opt
   private var options: ScoverageOptions = new ScoverageOptions()
   private var coverageFilter: CoverageFilter = AllCoverageFilter
 
+  // This [targetId] is used below in case [writeToClasspath] option
+  // is true. It is used for:
+  // 1. Creating a unique subdir under the classpath to store the instrument files, i.e
+  //    path to instrument files will be `< classpath >/META-INF/scoverage/< targetId >/scoverage.coverage`
+  // 2. Instrumenting the binaries with the [targetId], i.e binaries will be instrumented with
+  //    `Invoker.invokedWriteToClasspath(< statement id >, < targetId >)`
+  //    Consequently, at runtime, we can get the "correct" instrument file located at `.../< targetId >/scoverage.coverage`
+  //    from the classpath and store it with its appropriate measurements file.
+  private var targetId: String = ""
+
+  // used for generating the targetId.
+  def md5HashString(s: String): String = {
+    import java.security.MessageDigest
+    import java.math.BigInteger
+    val md = MessageDigest.getInstance("MD5")
+    val digest = md.digest(s.getBytes)
+    val bigInt = new BigInteger(1,digest)
+    val hashedString = bigInt.toString(16)
+    hashedString
+  }
+
   def setOptions(options: ScoverageOptions): Unit = {
+
     this.options = options
+
+    // Overriding the [dataDir] if [writeToClasspath] is true.
+    // If writeToClasspath is true, we set the output directory to output classpath,
+    // i.e to the directory where compiler is going to output the classfiles.
```

Comment:
This may not be unique, unfortunately. Even if you include the absolute path to the output directory, two compiles happening on different machines might use the same output directory, and not be identical.

I don't have any great ideas here. Hashing the source files themselves relative to the current working directory might work? Using a random number or UUID would make builds non-deterministic.

Maybe this information does need to be user provided, and you could require that the `dataDir` is a relative path that you'd use in this position?

Reply:
Agreed. I have currently set the `dataDir` to be the `targetID` in case the `writeToClasspath` option is set to true.

Reply:
Should update this comment, and below.